src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,245 @@
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Implementation of an asynchronous writer pool.
+ *
+ * @author Seth Stadick
+ */
+public class AsyncWriterPool implements Closeable {
+    private final ExecutorService executor;
+    private final AtomicBoolean poolClosed;
+    private final List<PooledWriter<?>> writers;
+
+    /**
+     * Create an AsyncWriterPool using the specified number of {@code threads}. The number of {@code threads} in use at
+     * one time will grow and shrink with load.
+     *
+     * @param threads max number of threads to use
+     */
+    public AsyncWriterPool(final int threads) {
+        assert threads >= 1; // TODO: throw error
+
+        this.poolClosed = new AtomicBoolean(false);
+        this.writers = new ArrayList<>();
+
+        this.executor = Executors.newWorkStealingPool(threads);
+    }
+
+    /**
+     * Create an AsyncWriterPool using all available processors. The number of threads in use at one time will grow and
+     * shrink with load.
+     */
+    public AsyncWriterPool() {
+        this(Runtime.getRuntime().availableProcessors());
+    }
+
+    /**
+     * Close all the pool and all writers in the pool.
+     *
+     * @throws IOException if any writer raises an exception
+     */
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        for (final PooledWriter<?> writer : this.writers) {
+            writer.close();
+        }
+        this.executor.shutdown();
+    }
+
+
+    /**
+     * Any class that implements {@link Writer} can be exchanged for a {@code PooledWriter}. The PooledWriter provides
+     * the same API as {@link Writer}, but will manage buffering of writes and sending to the {@link AsyncWriterPool} it
+     * is attached to.
+     * <p>
+     * Note, {@code PooledWriter} is not intrinsically threadsafe, and any writing / closing done across multiple
+     * threads should be synchronized.
+     *
+     * @param <A> the type of the item the writer can write
+     */
+    public class PooledWriter<A> implements Writer<A> {
+        private final Writer<A> writer;
+        private final BlockingQueue<A> queue;
+        private final AtomicBoolean writing;
+        private Boolean isClosed;
+
+        private int bufferSize;
+
+        // Holds the Future of the last task submitted to the AsyncWriterPools executor, or null if nothing has been
+        // submitted yet. Inside the Future is the number of items written and any exceptions that may have been thrown.
+        private Future<Void> currentTask;
+
+
+        /**
+         * Exchange a class implementing {@link Writer} for a {@code PooledWriter}.
+         *
+         * @param writer     a class implementing {@link Writer}
+         * @param queue      a queue to use for this writer, bounded or unbounded.
+         * @param bufferSize the minimum number of items to write needed before sending to a thread for writing.
```

Comment:
This sentence doesnt make size "to write needed"
```suggestion
         * @param bufferSize the minimum number of items needed before scheduling a thread for writing.
```